docs(galloc): introduce `dlmalloc` & `galloc` optimizations paper (#2645)
gear-tech/gear: galloc/docs/optimisation.md

# Optimisation of galloc
We've tried to optimise galloc in some ways, but none of them showed significant performance improvement, so we'd **strongly suggest to keep the current implementation.**

The current implementation uses static buffer in addition to standard <kbd>dlmalloc</kbd>-like implementation.

> ![Total perfomance comparison](./images/total-performance-comparison.png)

Author: Most of this work involved investigation and research, with some portions being done manually. However, I am uncertain about the necessity of uploading the scripts (since it's not the complete portion of the work), as well as generating charts in CI, given that the allocator is not frequently updated. What are your thoughts?

Here are some of the optimisation attempts we've tried:

1. [Measuring current performance](#current)
1. [Remove static buffer](#remove)
1. [Increase static buffer size by two](#increase)
1. [Multiple static buffers](#multiple)

## 0. Measuring current performance

> <a name="current"></a>

Author: Fixed, thanks

> ### How we measured

Author: Gas consumption will be the same for every machine, right?

> ### How we measured

Author: yes, definitely that needed additional explanation, thanks! I've added

> # Optimisation of galloc

Reviewer: ```suggestion
# Optimization of galloc
```

> # Optimization of galloc

Reviewer 2: Galloc implementations comparison

> We've tried to optimise galloc in some ways, but none of them showed significant performance improvement, so we'd **strongly suggest to keep the current implementation.**

Reviewer: ```suggestion
We've tried to optimize galloc in some ways, but none of them showed significant performance improvement, so we'd **strongly suggest keeping the current implementation.**
```

> We've tried to optimize galloc in some ways, but none of them showed significant performance improvement, so we'd **strongly suggest keeping the current implementation.**

Reviewer 2: This must be after `Content` title  in our case. In content this part must be named as `Summary` for developers who does not want to spent their time for detailed investigation.

> The current implementation uses static buffer in addition to standard <kbd>dlmalloc</kbd>-like implementation.

Reviewer: ```suggestion
The current implementation uses a static buffer in addition to standard <kbd>dlmalloc</kbd>-like implementation.
```

> ![Total perfomance comparison](./images/total-performance-comparison.png)

Reviewer 2: Add this as part of Content

Here are some of the optimization attempts we've tried:

> 

Reviewer 2: add title `Content`

> 1. [Measuring current performance](#current)

Reviewer 2: ```suggestion
1. [Current implementation](#current)
```

> 1. [Remove static buffer](#remove)

Reviewer 2: Implementation without static buffer

> 1. [Increase static buffer size by two](#increase)

Reviewer 2: Implementation with increased by two static buffer

> 1. [Multiple static buffers](#multiple)

Reviewer 2: Implementations with multiple static buffers

> ## 0. Measuring current performance

Reviewer 2: Please make numeration from `1` ))

<a name="current"></a>
We've measured the current [(commit 9135baa)](https://github.com/gear-tech/dlmalloc-rust/tree/9135baa728ef9a9a04a887998e019733c4b093af) performance of dlmalloc to have a baseline for comparison.

> ### How we measured

Reviewer 2: This must be separate part in `Content`. My suggestion name it `Benchmark`

All optimisation attempts were tested 3 times on the different test cases, and the average of the results was used to compare the performance. Our main metric is gas consumption: the less gas is consumed, the better. All optimisations were measured in release mode binaries with most optimisations enabled and compared to current perfomance.

The gas measurements were made via `debug` and `gas_available` syscalls, which was temporarily made free. Each allocation top-level operation (i.e. malloc, calloc, ...) was measured separately, the gas was measured before and after function call and calculated as substraction result of gas before and gas after.

> Then the gas consumption was calculated as a percentage of total gas spent on the test case.

Reviewer: ```suggestion
Then the gas consumption was calculated as a percentage of the total gas spent on the test case.
```

> We've intentionally did not put any info about testing machine, because we measure only gas consumption, which will be same for every machine, since we have fixed weights for every instruction and syscall.

Reviewer: ```suggestion
We've intentionally not put any info about the testing machine because we measure only gas consumption, which will be the same for every machine since we have fixed weights for every instruction and syscall.
```

The test cases were:

> - `NFT init -> mint -> burn`: This test case is for measuring the performance of the NFT contract, which is the one of the common cases of smart contracts. The test case consists of the following steps:

Reviewer: ```suggestion
- `NFT init -> mint -> burn`: This test case is for measuring the performance of the NFT contract, which is one of the common cases of smart contracts. The test case consists of the following steps:
```

  - `init`: Initialise the NFT contract.
  - `mint`: Mint 1 NFT.
  - `burn`: Burn 1 NFT.

>   This is quite simple case and it's not the most common case, but it's the most simple case we can think of.

Reviewer: ```suggestion
  This is quite a simple case, and it's not the most common case, but it's the simplest case we can think of.
```

  You can see the code of the test case [here](https://github.com/gear-dapps/non-fungible-token/blob/0.2.10/tests/node_tests.rs) (`burn-test`).

> - `FT stress-test` with <kbd>gtest</kbd>: This test case is for measuring the performance of the FT contract, which is also the one of the common cases of smart contracts. The test case consists of the following steps:

Reviewer: ```suggestion
- `FT stress-test` with <kbd>gtest</kbd>: This test case is for measuring the performance of the FT contract, which is also one of the common cases of smart contracts. The test case consists of the following steps:
```

  - `init`: Initialise the FT contract.
  - `mint`: Mint 1 000 000 FT to the first user.

>   - `transfer`: Transfer 6 000 FT to the first 20 accounts from first user.

Reviewer: ```suggestion
  - `transfer`: Transfer 6 000 FT to the first 20 accounts from the first user.
```

>   - `balance`: Check the balance of first account to prove it sent 6 000 FT to the first 20 accounts.

Reviewer: ```suggestion
  - `balance`: Check the balance of the first account to prove it sent 6 000 FT to the first 20 accounts.
```

>   - `balance`: Check the balance of first 20 accounts.

Reviewer: ```suggestion
  - `balance`: Check the balance of the first 20 accounts.
```

  - `transfer`: Transfer 1 000 FT from the first 20 accounts to users 21-100.

>   - `balance`: Check the balance of first 20 accounts to prove it sent 1 000 FT to users 21-100.

Reviewer: ```suggestion
  - `balance`: Check the balance of the first 20 accounts to prove it sent 1 000 FT to users 21-100.
```

  - `balance`: Check the balance of users 21-100 to prove they've received 1 000 FT from the first 20 accounts.
  - `mint`: Mint 1 000 000 FT to the second user. 
  - `mint`: Mint 5 000 FT to users 87..120.
  - `balance`: Check the balance of users 87..120 to prove they've received 5 000 FT after mint.
  - `transfer`: Transfer 1 000 FT from users 87..120 to the first user.
  - `total_supply`: Check the total supply of FT to prove it's sums up to correct value.
  - `balance`: Check the balance of users 918..13400
  - `mint`: Mint 5 000 FT to users 918..13400.
  - `transfer`: Transfer `i / 4` FT from users 918..13400 to the user `i * 2`.

  The next two steps were repeated 30 times:
  - `balance`: Check the balance of users 1..130
  - `transfer`: Transfer 1 FT from users 1..130 to the first user.

  This test tries to imitate what happens when the FT contract is used in the real world, with many clients and memory allocations therefore.

> - `FT stress-test` with <kbd>gclient</kbd>: This test case has the same steps as the previous one, but it's executed with <kbd>gclient</kbd> instead of <kbd>gtest</kbd>. This test is most similar to the real world case, because it's executed with the real node and will give the most real gas usage. 

Reviewer: ```suggestion
- `FT stress-test` with <kbd>gclient</kbd>: This test case has the same steps as the previous one, but it's executed with <kbd>gclient</kbd> instead of <kbd>gtest</kbd>. This test is most similar to the real world case because it's executed with the real node and will give the most realistic gas usage. 
```

#### `FT stress-test` with <kbd>gclient</kbd>
![FT stress-test with gclient](./images/average-cost-ft-gclient-current.png)
|  | malloc             | calloc            | realloc           | free                |
| -------------------------- | ------------------ | ----------------- | ----------------- | ------------------- |
| avg                        | 7 328 894          | 9 959 374         | 7 360 643         | 12 880 659          |
| min                        | 6 771 353          | 7 178 085         | 7 116 108         | 6 605 797           |
| max                        | 167 056 786        | 101 005 505       | 7 360 980         | 43 948 275          |
| count                      | 447 300            | 445 700           | 993 400           | 892 700             |
| total                      | 3 278 214 286 200  | 4 438 892 991 800 | 7 312 062 756 200 | 11 498 564 289 300  |
| % of total gas spent | 0,81%              | 1,10%             | 1,82%             | 2,86%               |

**Total gas spent:** 402 261 823 754 900

**Gas spent on allocations:** 26 527 734 323 500

[Go to top](#)

## 1. Remove static buffer
<a name="remove"></a>

> We've tried to remove static buffer and use only <kbd>dlmalloc</kbd>-like implementation, but it showed no significant performance improvement. In most cases, it even showed worse performance. So it seems that the static buffer is not the bottleneck.

Reviewer: ```suggestion
We've tried to remove the static buffer and use only <kbd>dlmalloc</kbd>-like implementation, but it showed no significant performance improvement. In most cases, it even showed worse performance. So it seems that the static buffer is not the bottleneck.
```

![NFT init -> mint -> burn](./images/average-cost-nft-no-static.png)

|| malloc        | calloc      | realloc          | free           |
| -------------------------- | ------------- | ----------- | ---------------- | -------------- |
| avg                        | 7 345 998     | 23 228 658  | 5 836 735        | 2 204 091      |
| min                        | 0             | 0           | 0                | 0              |
| max                        | 67 006 379    | 151 833 811 | 31 601 983       | 7 311 238      |
| count                      | 72            | 30          | 27               | 51             |
| total                      | 528 911 832   | 696 859 732 | 157 591 848      | 112 408 659    |
| % of total gas spent | 3,62%         | 4,77%       | 1,08%            | 0,77%          |

**Total gas spent:** 14 609 523 436

**Gas spent on allocations:** 1 495 772 071

#### `FT stress-test` with <kbd>gtest</kbd>
![FT stress-test with gtest](./images/average-cost-ft-gtest-no-static.png)

| | malloc        | calloc      | realloc          | free            |
| -------------------------- | ------------- | ----------- | ---------------- | --------------- |

> | avg                        | 3 633         | 10 305      | 4 184            | 3 672           |

Reviewer 2: Is not a good practice to show "absolute" values. Percents raw is enough. So, please delete avg, min, max and etc.

#### `FT stress-test` with <kbd>gclient</kbd>
![FT stress-test with gclient](./images/average-cost-ft-gclient-no-static.png)

| | malloc             | calloc            | realloc           | free                |
| -------------------------- | ------------------ | ----------------- | ----------------- | ------------------- |
| avg                        | 6 594 974          | 9 874 272         | 8 126 747         | 12 842 147          |
| min                        | 6 301 422          | 7 572 938         | 6 399 106         | 6 605 797           |
| max                        | 166 930 089        | 100 884 991       | 55 643 720        | 43 948 275          |
| count                      | 447 300            | 445 700           | 993 400           | 892 700             |
| total                      | 2 949 931 870 200  | 4 400 963 030 400 | 8 073 110 469 800 | 11 464 184 626 900  |
| % of total gas spent | 0,73%              | 1,09%             | 2,01%             | 2,85%               |

**Total gas spent:** 401 929 928 086 200

**Gas spent on allocations:** 26 888 189 997 300
[Go to top](#)

## 2. Increase static buffer size by two
<a name="increase"></a>

> We've tried to increase static buffer size by two, and it's perfomance was slightly better than the current implementation. But it's not significant enough to change the current implementation.

Reviewer: ```suggestion
We've tried to increase the static buffer size by two, and its performance was slightly better than the current implementation. But it's not significant enough to change the current implementation.
```

#### `FT stress-test` with <kbd>gclient</kbd>

![FT stress-test with gclient](./images/average-cost-ft-gclient-increase-twice.png)

| | malloc             | calloc            | realloc           | free                |
| -------------------------- | ------------------ | ----------------- | ----------------- | ------------------- |
| avg                        | 7 328 883          | 9 959 385         | 7 360 629         | 12 880 657          |
| min                        | 6 721 051          | 7 129 901         | 6 978 562         | 6 605 797           |
| max                        | 167 056 786        | 101 005 505       | 7 360 980         | 43 948 275          |
| count                      | 447 300            | 445 700           | 993 400           | 892 700             |
| total                      | 3 278 209 365 900  | 4 438 897 894 500 | 7 312 048 848 600 | 11 498 562 503 900  |
| % of total gas spent | 0,82%              | 1,10%             | 1,82%             | 2,86%               |

**Total gas spent:** 402 127 166 124 400

**Gas spent on allocations:** 26 527 718 612 900

[Go to top](#)

## 3. Multiple static buffers
<a name="multiple"></a>

> We've tried to implement not only one static buffer, but multiple static buffers, with different sizes. But the study showed that it's not worth the effort and most allocations was done in the largest static buffer, which perfomance was even worster to the current implementation.

Reviewer: ```suggestion
We've tried to implement not only one static buffer but multiple static buffers with different sizes. But the study showed that it's not worth the effort, and most allocations were done in the largest static buffer, which performance was even worse than the current implementation.
```
